File: pact-jvm-matchers/src/main/scala/au/com/dius/pact/matchers/Matchers.scala
package au.com.dius.pact.matchers

import java.util.function.{Predicate, ToIntFunction}

import au.com.dius.pact.model.matchingrules.{MatchingRule, MatchingRules}
import com.typesafe.scalalogging.StrictLogging
import io.gatling.jsonpath.AST._
import io.gatling.jsonpath.Parser

import scala.collection.JavaConversions._

object Matchers extends StrictLogging {

  def matchesToken(pathElement: String, token: PathToken) = token match {
    case RootNode => if (pathElement == "$") 2 else 0
    case Field(name) => if (pathElement == name) 2 else 0
    case ArrayRandomAccess(indices) => if (pathElement.matches("\\d+") && indices.contains(pathElement.toInt)) 2 else 0
    case ArraySlice(None, None, 1) => if (pathElement.matches("\\d+")) 1 else 0
    case AnyField => 1
    case _ => 0
  }

  def matchesPath(pathExp: String, path: Seq[String]) =
    new Parser().compile(pathExp) match {
      case Parser.Success(q, _) =>
        val filter = path.reverse.tails.filter(l =>
          l.reverse.corresponds(q)((pathElement, pathToken) => matchesToken(pathElement, pathToken) != 0))
        if (filter.nonEmpty) {
          filter.maxBy(seq => seq.length).length
        } else {
          0
        }
      case ns: Parser.NoSuccess =>
        logger.warn(s"Path expression $pathExp is invalid, ignoring: $ns")
        0
    }

  def calculatePathWeight(pathExp: String, path: Seq[String]) = {
    new Parser().compile(pathExp) match {
      case Parser.Success(q, _) =>
        path.zip(q).map(entry => matchesToken(entry._1, entry._2)).product
      case ns: Parser.NoSuccess =>
        logger.warn(s"Path expression $pathExp is invalid, ignoring: $ns")
        0
    }
  }

  def resolveMatchers(matchers: MatchingRules, category: String, items: Seq[String]) =
    if (category == "body")
      matchers.rulesForCategory(category).filter(new Predicate[String] {
        override def test(p: String): Boolean = matchesPath(p, items) > 0
      })
    else if (category == "header" || category == "query")
      matchers.rulesForCategory(category).filter(new Predicate[String] {
        override def test(p: String): Boolean = items == Seq(p)
      })
    else matchers.rulesForCategory(category)

  def matcherDefined(category: String, path: Seq[String], matchers: MatchingRules): Boolean =
    if (matchers != null)
      resolveMatchers(matchers, category, path).isNotEmpty
    else
      false

  def wildcardMatcherDefined(path: Seq[String], category: String, matchers: MatchingRules): Boolean = {
    if (matchers != null) {
      val resolvedMatchers = matchers.rulesForCategory(category).filter(new Predicate[String] {
        override def test(p: String): Boolean = matchesPath(p, path) == path.length
      })
      asScalaSet(resolvedMatchers.getMatchingRules.keySet()).exists(entry => entry.endsWith(".*"))
    } else
      false
  }

  def domatch[Mismatch](matchers: MatchingRules, category: String, path: Seq[String], expected: Any, actual: Any,
                        mismatchFn: MismatchFactory[Mismatch]) : List[Mismatch] = {
    val matcherDef = selectBestMatcher(matchers, category, path)
    asScalaBuffer(MatcherExecutorKt.domatch(matcherDef, path, expected, actual, mismatchFn)).toList
  }

  def selectBestMatcher[Mismatch](matchers: MatchingRules, category: String, path: Seq[String]) = {
    val matcherCategory = resolveMatchers(matchers, category, path)
    if (category == "body")
      matcherCategory.maxBy(new ToIntFunction[String] {
        override def applyAsInt(value: String): Int = calculatePathWeight(value, path)
      })
    else
      matcherCategory.getMatchingRules.iterator.next()._2
  }
}

File: pact-jvm-matchers/src/main/kotlin/au/com/dius/pact/matchers/Matchers.kt
package au.com.dius.pact.matchers

import au.com.dius.pact.matchers.util.corresponds
import au.com.dius.pact.matchers.util.tails
import au.com.dius.pact.model.matchingrules.MatchingRuleGroup
import au.com.dius.pact.model.matchingrules.MatchingRules
import io.gatling.jsonpath.AST
import io.gatling.jsonpath.Parser
import mu.KLogging
import scala.collection.JavaConversions
import scala.collection.JavaConverters
import scala.util.parsing.combinator.Parsers
import java.util.Comparator
import java.util.function.Predicate

object Matchers : KLogging() {

  const val PACT_MATCHING_WILDCARD = "pact.matching.wildcard"

  fun matchesToken(pathElement: String, token: AST.PathToken) = when (token) {
    is AST.`RootNode$` -> if (pathElement == "$") 2 else 0
    is AST.Field -> if (pathElement == token.name()) 2 else 0
    is AST.ArrayRandomAccess -> if (pathElement.matches(Regex("\\d+")) && token.indices().contains(pathElement.toInt())) 2 else 0
    is AST.ArraySlice -> if (pathElement.matches(Regex("\\d+"))) 1 else 0
    is AST.`AnyField$` -> 1
    else -> 0
  }

  fun matchesPath(pathExp: String, path: List<String>): Int {
    val parseResult = Parser().compile(pathExp)
    return when (parseResult) {
      is Parsers.Success -> {
        val filter = tails(path.reversed()).filter { l ->
          corresponds(l.reversed(), JavaConversions.asJavaCollection(parseResult.result()).toList()) { pathElement, pathToken ->
            matchesToken(pathElement, pathToken) != 0
          }
        }
        if (filter.isNotEmpty()) {
          filter.maxBy { seq -> seq.size }?.size ?: 0
        } else {
          0
        }
      }
      else -> {
        logger.warn { "Path expression $pathExp is invalid, ignoring: $parseResult" }
        0
      }
    }
  }

  fun calculatePathWeight(pathExp: String, path: List<String>): Int {
    val parseResult = Parser().compile(pathExp)
    return when (parseResult) {
      is Parsers.Success -> path.zip(JavaConverters.asJavaCollectionConverter(parseResult.result()).asJavaCollection()).map {
        matchesToken(it.first, it.second)
      }.reduce { acc, i -> acc * i }
      else -> {
        logger.warn { "Path expression $pathExp is invalid, ignoring: $parseResult" }
        0
      }
    }
  }

  fun resolveMatchers(
    matchers: MatchingRules,
    category: String,
    items: List<String>,
    pathComparator: Comparator<String>
  ) = if (category == "body")
      matchers.rulesForCategory(category).filter(Predicate {
        matchesPath(it, items) > 0
      })
    else if (category == "header" || category == "query")
      matchers.rulesForCategory(category).filter(Predicate { key ->
        items.all { pathComparator.compare(key, it) == 0 }
      })
    else matchers.rulesForCategory(category)

  @JvmStatic
  @JvmOverloads
  fun matcherDefined(
    category: String,
    path: List<String>,
    matchers: MatchingRules?,
    pathComparator: Comparator<String> = Comparator.naturalOrder()
  ): Boolean =
    if (matchers != null)
      resolveMatchers(matchers, category, path, pathComparator).isNotEmpty()
    else false

  /**
   * Determines if a matcher of the form '.*' exists for the path
   */
  @JvmStatic
  fun wildcardMatcherDefined(path: List<String>, category: String, matchers: MatchingRules?) =
    if (matchers != null) {
      val resolvedMatchers = matchers.rulesForCategory(category).filter(Predicate {
        matchesPath(it, path) == path.size
      })
      resolvedMatchers.matchingRules.keys.any { entry -> entry.endsWith(".*") }
    } else {
      false
    }

  /**
   * If wildcard matching logic is enabled (where keys are ignored and only values are compared)
   */
  @JvmStatic
  fun wildcardMatchingEnabled() = System.getProperty(PACT_MATCHING_WILDCARD)?.trim() == "true"

  @JvmStatic
  @JvmOverloads
  fun <M : Mismatch> domatch(
    matchers: MatchingRules,
    category: String,
    path: List<String>,
    expected: Any?,
    actual: Any?,
    mismatchFn: MismatchFactory<M>,
    pathComparator: Comparator<String> = Comparator.naturalOrder()
  ): List<M> {
    val matcherDef = selectBestMatcher(matchers, category, path, pathComparator)
    return domatch(matcherDef, path, expected, actual, mismatchFn)
  }

  @JvmStatic
  @JvmOverloads
  fun selectBestMatcher(
    matchers: MatchingRules,
    category: String,
    path: List<String>,
    pathComparator: Comparator<String> = Comparator.naturalOrder()
  ): MatchingRuleGroup {
    val matcherCategory = resolveMatchers(matchers, category, path, pathComparator)
    return if (category == "body")
      matcherCategory.maxBy(Comparator { a, b ->
        val weightA = calculatePathWeight(a, path)
        val weightB = calculatePathWeight(b, path)
        when {
          weightA == weightB -> when {
            a.length > b.length -> 1
            a.length < b.length -> -1
            else -> 0
          }
          weightA > weightB -> 1
          else -> -1
        }
      })
    else {
      matcherCategory.matchingRules.values.first()
    }
  }
}
